tests/scanner/audit/retention_rules_engine_test.py

```diff
@@ -0,0 +1,767 @@
+
+    yaml_str_only_min_retention = """
+rules:
+  - name: only min retention
+    applies_to:
+      - bucket
+    resource:
+      - type: bucket
+        resource_ids:
+          - fake_bucket
+    minimum_retention: 90
+
+"""
+
+    def test_only_min_normal_del(self):
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_only_min_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+        data_creater.AddLefecycleDict(action="Delete", age=90)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        self.assertEqual(got_violations, [])
+
+    def test_only_min_normal_else(self):
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_only_min_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+        data_creater.AddLefecycleDict(action="SetStorageClass", age=90)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        self.assertEqual(got_violations, [])
+
+    def test_only_min_less_else(self):
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_only_min_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+        data_creater.AddLefecycleDict(action="SetStorageClass", age=89)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        self.assertEqual(got_violations, [])
+
+    def test_only_min_no_condition(self):
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_only_min_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        self.assertEqual(got_violations, [])
+
+    def test_only_min_lessver1_del(self):
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_only_min_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+        data_creater.AddLefecycleDict(action="Delete", age=89, num_newer_versions=1)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        self.assertEqual(got_violations, [])
+
+    def test_only_min_lessver0_del(self):
```

Comment:
Please add a comment/docstring to each of these tests to explain what they are testing.